PSA Internal Functions being included even with MBEDTLS_PSA_CRYPTO_C undefined

I'm trying to compile mbedTLS 2.26.0 for Windows using CMake, directly on Windows (build=host=Windows). GCC is version 10.3 and checked the option USE_SHARED_MBEDTLS_LIBRARY.

I get the following errors in the linking step:

```
[ 76%] Linking C shared library libmbedcrypto.dll
c:/mingw/bin/../lib/gcc/i686-w64-mingw32/10.3.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\mbedcrypto.dir/objects.a(psa_crypto_driver_wrappers.c.obj):psa_crypto_driver_wrappers.c:(.text+0x33): undefined reference to `psa_sign_hash_internal'
c:/mingw/bin/../lib/gcc/i686-w64-mingw32/10.3.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\mbedcrypto.dir/objects.a(psa_crypto_driver_wrappers.c.obj):psa_crypto_driver_wrappers.c:(.text+0x7b): undefined reference to `psa_verify_hash_internal'
c:/mingw/bin/../lib/gcc/i686-w64-mingw32/10.3.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\mbedcrypto.dir/objects.a(psa_crypto_driver_wrappers.c.obj):psa_crypto_driver_wrappers.c:(.text+0xb1): undefined reference to `psa_generate_key_internal'
c:/mingw/bin/../lib/gcc/i686-w64-mingw32/10.3.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\mbedcrypto.dir/objects.a(psa_crypto_driver_wrappers.c.obj):psa_crypto_driver_wrappers.c:(.text+0xe3): undefined reference to `psa_import_key_into_slot'
c:/mingw/bin/../lib/gcc/i686-w64-mingw32/10.3.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\mbedcrypto.dir/objects.a(psa_crypto_driver_wrappers.c.obj):psa_crypto_driver_wrappers.c:(.text+0x115): undefined reference to `psa_export_key_internal'
c:/mingw/bin/../lib/gcc/i686-w64-mingw32/10.3.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\mbedcrypto.dir/objects.a(psa_crypto_driver_wrappers.c.obj):psa_crypto_driver_wrappers.c:(.text+0x145): undefined reference to `psa_export_public_key_internal'
```

My config file has the following:

```c
#define MBEDTLS_HAVE_TIME
#define MBEDTLS_PLATFORM_C

#define MBEDTLS_REMOVE_ARC4_CIPHERSUITES

#define MBEDTLS_RSA_C
#define MBEDTLS_BIGNUM_C
#define MBEDTLS_PK_C
#define MBEDTLS_PK_PARSE_C
#define MBEDTLS_PK_WRITE_C
#define MBEDTLS_PK_RSA_ALT_SUPPORT
#define MBEDTLS_PKCS1_V15
#define MBEDTLS_ASN1_PARSE_C
#define MBEDTLS_ASN1_WRITE_C
#define MBEDTLS_PEM_PARSE_C

#define MBEDTLS_CIPHER_C
#define MBEDTLS_CIPHER_MODE_CBC
#define MBEDTLS_AES_C
#define MBEDTLS_DES_C
#define MBEDTLS_HMAC_DRBG_C
#define MBEDTLS_OID_C

#define MBEDTLS_MD_C
#define MBEDTLS_SHA1_C
#define MBEDTLS_SHA256_C

#define MBEDTLS_BASE64_C
#define MBEDTLS_ERROR_C
#define MBEDTLS_VERSION_C
```

In my IDE, I see *psa_crypto.c* properly greyed out, where the implementations of functions like psa_sign_hash_internal are. But in **psa_crypto_driver_wrappers.c:112** the code returns psa_sign_hash_internal, causing the undefined reference since its implementation is not included in compilation. The same applies for the other functions in the error log.

I even added

```c
#undef MBEDTLS_PSA_CRYPTO_C
#undef MBEDTLS_PSA_CRYPTO_DRIVERS
```

Am I missing something?

--- a/ChangeLog.d/undefined_reference_without_psa.txt
+++ b/ChangeLog.d/undefined_reference_without_psa.txt
@@ -1,0 +1,4 @@
+Bugfix
+   * With MBEDTLS_PSA_CRYPTO_C disabled, some functions were getting built
+     nonetheless, resulting in undefined reference errors when building a
+     shared library. Reported by Guillermo Garcia M. in #4411.

--- a/library/psa_crypto_driver_wrappers.c
+++ b/library/psa_crypto_driver_wrappers.c
@@ -28,6 +28,8 @@
 
 #include "mbedtls/platform.h"
 
+#if defined(MBEDTLS_PSA_CRYPTO_C)
+
 #if defined(MBEDTLS_PSA_CRYPTO_DRIVERS)
 
 /* Include test driver definition when running tests */
@@ -1777,4 +1779,5 @@ psa_status_t psa_driver_wrapper_mac_abort(
             return( PSA_ERROR_INVALID_ARGUMENT );
     }
 }
-/* End of automatically generated file. */
+
+#endif /* MBEDTLS_PSA_CRYPTO_C */

